inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp

```diff
@@ -2082,22 +2080,41 @@ void MKLDNNGraphOptimizer::MergePermuteAndReorder(MKLDNNGraph &graph) {
         return true;
     };
 
-    // Permute and Reorder do opposite permutation to each other.
+    // Transpose and Reorder do opposite permutation to each other.
     // Example:
-    //      chain [physical layout: NCHW, logical layout: NCHW] -> Permute(order=0312) -> [physical layout: NWCH, logical layout: NCHW] ->
+    //      chain [physical layout: NCHW, logical layout: NCHW] -> Transpose(order=0312) -> [physical layout: NWCH, logical layout: NCHW] ->
     //      Reorder(nchw->nhwc) -> [physical layout: NCHW, logical layout: NHWC] can be replaced with Reorder(nchw->nhwc; isOptimized=true)
     //      which will just reinterprets layout without physical change of the memory.
     // Two cases are possible:
     //      1) inPrec = outPrec
-    //          In this case, we replace Permute+Reorder pattern with a new Reorder that does nothing.
+    //          In this case, we replace Transpose+Reorder pattern with a new Reorder that does nothing.
     //      2) inPrec != outPrec
-    //          As in the first case, we also replace Permute+Reorder pattern with a new Reorder.
+    //          As in the first case, we also replace Transpose+Reorder pattern with a new Reorder.
     //          Additionally, we insert another Reorder that performs the conversion from the input precision (inPrec)
     //          to the output precision (outPrec)
-    auto mergePermuteAndReorder = [&](std::shared_ptr<MKLDNNNode>& parentNode, std::shared_ptr<MKLDNNNode>& childNode) {
-        auto parentParentNode = parentNode->getParentEdgeAt(0)->getParent();
+    auto mergeTransposeAndReorder = [&](std::shared_ptr<MKLDNNNode>& parentNode, std::shared_ptr<MKLDNNNode>& childNode) {
+        MKLDNNNodePtr parentParentNode;
+        MKLDNNNodePtr parentParentConstNode;
+        if (parentNode->getParentEdgeAt(1)->getParent()->isConstant()) {
+            parentParentNode = parentNode->getParentEdgeAt(0)->getParent();
+            parentParentConstNode = parentNode->getParentEdgeAt(1)->getParent();
+        } else {
+            parentParentNode = parentNode->getParentEdgeAt(1)->getParent();
+            parentParentConstNode = parentNode->getParentEdgeAt(0)->getParent(); // TODO: why const can be on the 0 edge?
```

Comment:
Because topological sort can be broken during previous optimization. You need to use getParentEdgesAtPort() to get correct input.

Reply:
Fixed